tests/test_typing.py

```diff
@@ -0,0 +1,158 @@
+        ),
+        param(
+            ListConfig([User], element_type=User),
+            0,
+            None,
+            True,
+            id="Optional[User]<-str_fails",
+        ),
+    ],
+)
+def test_assign(cls: Any, key: str, assignment: Any, error: bool) -> None:
+    cfg = OmegaConf.structured(cls)
+    if error:
+        with raises(ValidationError):
+            cfg[key] = assignment
+    else:
+        cfg[key] = assignment
+        assert cfg[key] == assignment
+
+
+@mark.parametrize(
+    "src,op,keys,ref_type,is_optional",
+    [
+        param(Group, None, ["admin"], User, True, id="opt_user"),
+        param(
+            ConcretePlugin,
+            None,
+            ["params"],
+            ConcretePlugin.FoobarParams,
+            False,
+            id="nested_structured_conf",
+        ),
+        param(
+            OmegaConf.structured(Users({"user007": User("Bond", 7)})).name2user,
+            None,
+            ["user007"],
+            User,
+            False,
+            id="structured_dict_of_user",
+        ),
+        param(
+            DictConfig({"a": 123}, element_type=int),
+            None,
+            ["a"],
+            int,
+            False,
+            id="dict_int",
+        ),
+        param(
+            DictConfig({"a": 123}, element_type=Optional[int]),
+            None,
+            ["a"],
+            int,
+            True,
+            id="dict_opt_int",
+        ),
+        param(DictConfig({"a": 123}), None, ["a"], Any, True, id="dict_any"),
+        param(
+            ListConfig([], element_type=int),
+            lambda cfg: cfg.insert(0, 123),
+            [0],
+            int,
+            False,
+            id="list_int_insert",
+        ),
+        param(
+            OmegaConf.merge(Users, {"name2user": {"joe": User("joe")}}),
+            None,
+            ["name2user", "joe"],
+            User,
+            False,
+            id="dict:merge_into_new_user_node",
+        ),
+        param(
+            OmegaConf.merge(OptionalUsers, {"name2user": {"joe": User("joe")}}),
+            None,
+            ["name2user", "joe"],
+            User,
+            True,
+            id="dict:merge_into_new_optional_user_node",
+        ),
+        param(
+            OmegaConf.merge(ListConfig([], element_type=User), [User(name="joe")]),
+            None,
+            [0],
+            User,
+            False,
+            id="list:merge_into_new_user_node",
+        ),
+        param(
+            OmegaConf.merge(
+                ListConfig([], element_type=Optional[User]), [User(name="joe")]
+            ),
+            None,
+            [0],
+            User,
+            True,
+            id="list:merge_into_new_optional_user_node",
+        ),
+    ],
+)
+def test_ref_type(
+    src: Any, op: Any, keys: Any, ref_type: Any, is_optional: bool
+) -> None:
+    cfg = _ensure_container(src)
+    if callable(op):
+        op(cfg)
+    for k in keys:
```

Comment:
Minor: I would suggest to accept a single key as well, i.e.:
```python
if not isinstance(keys, (list, tuple)):
    keys = [keys]
```
=> makes test cases simpler and avoids confusing errors when forgetting to put the string into a list

Reply:
Oh, nice! Done in https://github.com/omry/omegaconf/pull/749/commits/da7457330e40d5d166dae84e89bbb3d90f8b2cec.